File: Python/stack.py
class Stack(object):

    def __init__(self, length):
        self.__items = [None] * length
        self.__index = 0

    def __len__(self):
        return self.__index

    @property
    def is_empty(self):
        return self.__index == 0

    @property
    def is_full(self):
        return self.__index == len(self.__items)

    def peek(self):

        if self.is_empty:
            raise IndexError("Cannot peek at an empty stack!")

        return self.__items[self.__index - 1]

    def pop(self):

        if self.is_empty:
            raise IndexError("Cannot pop from an empty stack!")

        self.__index -= 1
        item = self.__items[self.__index]
        self.__items[self.__index] = None
        return item

    def push(self, item):

        if self.is_full:
            raise IndexError("Cannot push into a full stack!")

        self.__items[self.__index] = item
        self.__index += 1

File: Kotlin/stack.kt
internal class Stack<T>(size: Int) {

  internal var size: Int
    get() = this.index

  internal var isEmpty: Boolean = true
    get() = this.index == 0

  internal var isFull: Boolean = false
    get() = this.index == this.items.size

  private var items: Array<Any?>
  private var index: Int

  init {
    this.items = arrayOfNulls<Any?>(size)
    this.index = 0
  }

  internal fun peek(): T? {

    if (this.isEmpty) {
      throw Exception("Cannot peek at an empty stack!")
    }

    return this.items[this.index - 1] as T?
  }

  internal fun pop(): T? {

    if (this.isEmpty) {
      throw Exception("Cannot pop from an empty stack!")
    }

    this.index -= 1
    val item = this.items[this.index]
    this.items[this.index] = null
    return item as T?
  }

  internal fun push(item: T?): Unit {

    if (this.isFull) {
      throw Exception("Cannot push into a full stack!")
    }

    this.items[this.index] = item
    this.index += 1
  }

}
